addons/website/models/website.py

```diff
@@ -241,11 +245,138 @@ def unlink(self):
+        r['industries'] = self._iap_website_service('/website/industries', {'lang': self.env.user.lang})['industries']
+        return r
+
+    @api.model
+    def configurator_recommended_themes(self, industry_name, palette):
+        domain = [('name', '=like', 'theme%'), ('name', 'not in', ['theme_default', 'theme_common'])]
+        client_themes = [theme['name'] for theme in request.env['ir.module.module'].search_read(domain, ['name'])]
+        params = {
+            'industry_name': industry_name,
+            'palette': palette,
+            'client_themes': client_themes,
+        }
+        return self._iap_website_service('/website/recommended_themes', params)
+
+    @api.model
+    def configurator_skip(self):
+        self.configurator_done = True
+
+    @api.model
+    def configurator_apply(self, data):
+        def set_colors(selected_palette):
+            url = '/website/static/src/scss/options/user_values.scss'
+            custo = {'color-palettes-name': "'%s'" % selected_palette}
+            self.env['web_editor.assets'].make_scss_customization(url, custo)
+
+        def set_features(selected_features):
+            feature_ids = self.env['website.configurator.feature'].browse(selected_features)
+            pages_views = {}
+            modules = self.env['ir.module.module']
+            for feature_id in feature_ids:
+                if feature_id.type == 'app' and feature_id.module_id and feature_id.module_id.state != 'installed':
+                    modules += feature_id.module_id
+                if feature_id.type == 'page' and feature_id.page_view_id:
+                    result = self.env['website'].new_page(name=feature_id.name, add_menu=True, template=feature_id.page_view_id.key)
+                    pages_views[feature_id.iap_page_code] = result['view_id']
+
+            modules._button_immediate_function(lambda self: self.button_install())
+            return pages_views
+
+        def configure_page(page_code, snippet_list, pages_views):
+            if page_code == 'homepage':
+                page_view_id = website.homepage_id.view_id
+            else:
+                page_view_id = self.env['ir.ui.view'].browse(pages_views[page_code])
+            page_snippets = b''
+            for snippet in snippet_list:
+                try:
+                    view_id = self.env['website'].with_context(website_id=website.id).viewref(snippet)
+                    if view_id:
+                        page_snippets += view_id._render()
+                except ValueError as e:
+                    logger.warning(e)
+            page_view_id.save(value=page_snippets.decode('utf-8'), xpath="(//div[hasclass('oe_structure')])[last()]")
+
+        def set_images(images):
+            for image in images:
+                try:
+                    response = requests.get(image['url'], timeout=3)
+                    response.raise_for_status()
+                    datas = base64.b64encode(response.content).decode('utf-8')
```

Comment:
utf-8 is implicit, but actually you don't need to play that base64 dance anymore, just use the `raw` field instead of `datas` and you can read/write the original binary value (response.content) directly